core/src/main/java/com/linecorp/armeria/common/stream/ByteBufDeframerInput.java

```diff
@@ -0,0 +1,219 @@
+    }
+
+    @Override
+    public int readInt() {
+        final ByteBuf firstBuf = queue.peek();
+        if (firstBuf == null) {
+            throw new IndexOutOfBoundsException(
+                    "readerIndex: " + readerIndex() +
+                    " (expected: 0 <= readerIndex < writerIndex(" + writerIndex() + "))");
+        }
+
+        if (firstBuf.readableBytes() >= 4) {
+            return firstBuf.readInt();
+        }
+
+        return readIntSlow();
+    }
+
+    private int readIntSlow() {
+        int value = 0;
+        int remained = 4;
+        for (ByteBuf buf : queue) {
+            if (!buf.isReadable()) {
+                continue;
+            }
+            final int readSize = Math.min(remained, buf.readableBytes());
+            if (readSize == 4) {
+                return buf.readInt();
+            }
+
+            value <<= 8 * readSize;
+            switch (readSize) {
+                case 1:
+                    value |= buf.readUnsignedByte();
+                    break;
+                case 2:
+                    value |= buf.readUnsignedShort();
+                    break;
+                case 3:
+                    value |= buf.readUnsignedMedium();
+                    break;
+                default:
+                    throw new Error(); // Should not reach here.
+            }
+            remained -= readSize;
+            if (remained == 0) {
+                break;
+            }
+        }
+
+        if (remained == 0) {
+            return value;
+        } else {
+            throw new IndexOutOfBoundsException(
+                    "readerIndex: " + (readerIndex() - 4 + remained) +
+                    " (expected: 0 < readerIndex + 4 <= writerIndex(" + writerIndex() + "))");
+        }
+    }
+
+    @Override
+    public ByteBuf readBytes(int length) {
+        checkArgument(length > 0, "length %s (expected: length > 0)", length);
+        final ByteBuf firstBuf = queue.peek();
+        if (firstBuf == null) {
+            throw new IndexOutOfBoundsException(
+                    "readerIndex: " + readerIndex() +
+                    " (expected: 0 <= readerIndex < writerIndex(" + writerIndex() + "))");
+        }
+
+        if (firstBuf.readableBytes() >= length) {
+            return firstBuf.readRetainedSlice(length);
+        }
+
+        return readBytesSlow(length);
+    }
+
+    private ByteBuf readBytesSlow(int length) {
+        ByteBuf value = null;
+        int remained = length;
+        for (ByteBuf buf : queue) {
+            if (!buf.isReadable()) {
+                continue;
+            }
+
+            final int readSize = Math.min(remained, buf.readableBytes());
+            if (readSize == length) {
+                return buf.readRetainedSlice(readSize);
+            } else {
+                if (value == null) {
+                    value = alloc.buffer(length);
+                }
+                value.writeBytes(buf, readSize);
```

Comment:
Question: do you think it's worth releasing the buf as soon as it's consumed?

Reply:
The consumed `buf` will be released right after `process()` is returned.
I think it would be nice to release `buf` only one time then checking every time whenever `buf` is read.